feat(cli): Status information about all backup profiles

s4moore opened the pull request:
Implemented the 'status' command to provide a snapshot of the status of all profiles. 

options:
  -h, --help            show this help message and exit
  --profile NAME        a more detailed summary of the profile with the name, NAME
  --profile-id ID       a more detailed summary of the profile with the id, ID
  --issues              show only profiles with errors on most recent run or no snapshot history at all
  --human-readable, -r  print in human readable format

Consider adding a CHANGELOG entry.
Run "codespell" to check for typos.

Close #1018

s4moore committed:
implement status command for summary of snapshots

s4moore committed:
remove extra line [skip ci]

s4moore committed:
fix typo [skip ci]

s4moore committed:
implement profile specific status [skip ci]

buhtz commented:
Looks nice in a first place. You are on the right track.

s4moore committed:
Merge branch 'dev' into snapshot_status

buhtz committed:
Merge branch 'dev' into snapshot_status

buhtz commented on common/backintime.py:
```diff
@@ -464,6 +464,36 @@ def createParsers(app_name='backintime'):
+    snapshotStatusCP =       subparsers.add_parser(command,
+                                                 epilog = epilogCommon,
+                                                 help = description,
+                                                 description = description)
```
Please do not copy the fancy coding style of previous developers. ;) Just stick to PEP8 please.

```suggestion
    snapshotStatusCP = subparsers.add_parser(command,
                                             epilog=epilogCommon,
                                             help=description,
                                             description=description)
```

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+            info.update({_('Last successful'):
+                str(last_success) if last_success else None})
+        return {config.profileName(config.currentProfile()): info}
+    except:
```
It is a [to broad exception](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/broad-exception-raised.html). Please catch a specific exception.

s4moore replied:
Okay, changed this slightly and made exception more specific

s4moore replied:
Done

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+            }})
+    print(json.dumps(info, indent=2), file=force_stdout)
+    _umount(cfg)
+    sys.exit(RETURN_OK)
```
Don't be scared of blank lines. ;)

```suggestion
    force_stdout = setQuiet(args)
    cfg = getConfig(args)
    _mount(cfg)

    if args.profile:
        if not cfg.setCurrentProfileByName(args.profile):
            _umount(cfg)
            logger.error(args.profile)

    elif not cfg.setCurrentProfile(args.profile_id):
        _umount(cfg)
        logger.error(args.profile_id)

    id = cfg.currentProfile()
    info = lastSnapshotDict(cfg)
    info[cfg.profileName(id)].update({
            _('Paths'): {
                _('Snapshots'): cfg.snapshotsFullPath(),
                _('Path exists'): cfg.canBackup(),
                _('Mount point'): cfg.get_snapshots_mountpoint(),
                _('Log file'): cfg.takeSnapshotLogFile(),
            }})

    print(json.dumps(info, indent=2), file=force_stdout)
    _umount(cfg)

    sys.exit(RETURN_OK)
```

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+    cfg = getConfig(args)
+    _mount(cfg)
+    status = {}
+    for profile in cfg.profiles():
```
```suggestion
    status = {}

    for profile in cfg.profiles():
```

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+    """
+    if args.profile or args.profile_id:
+        profileStatus(args)
+    force_stdout = setQuiet(args)
```
```suggestion
        profileStatus(args)

    force_stdout = setQuiet(args)
```

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+    for profile in cfg.profiles():
+        cfg.setCurrentProfile(profile)
+        profile_dict = lastSnapshotDict(cfg)
+        if not args.quiet or \
```
```suggestion
        profile_dict = lastSnapshotDict(cfg)

        if not args.quiet or \
```

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+        if not args.quiet or \
+            not profile_dict[cfg.profileName(profile)].get('Successful'):
+            status.update(profile_dict)
+    if not args.human_readable:
```
```suggestion
            status.update(profile_dict)

    if not args.human_readable:
```

buhtz commented on common/backintime.py:
```diff
@@ -1144,6 +1174,113 @@ def snapshotsListPath(args):
+    else:
+        printDictHumanReadable(status, force_stdout)
+    _umount(cfg)
+    sys.exit(RETURN_OK)
```
```suggestion
    _umount(cfg)

    sys.exit(RETURN_OK)
```

buhtz reviewed (changes requested):
Be aware that I updated your branch to the latest dev. So you have to update your local working branch, too.

- Please set the PR into Draft mode if it isn't finished yet. This helps me organizing.
- Added some PEP8 related suggestions.
- Feel free to add yourself into the SPDX header of backintime.py.
- Different from my first comments about it I would prefer to make human readable output the default behavior. So add `--json` as a new switch and remove `--human-readable`.

The `--issues` switch is not working as expected. I might misunderstand something?

```sh
$ ./backintime status --issues
usage: backintime status [-h] [--profile NAME | --profile-id ID | --issues ] [--human-readable]
backintime status: error: argument --issues: expected one argument

$ ./backintime status --profile 1 --issues
usage: backintime status [-h] [--profile NAME | --profile-id ID | --issues ] [--human-readable]
backintime status: error: argument --issues: expected one argument
```

buhtz committed:
Merge branch 'dev' into snapshot_status

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/backintime.py [skip ci]

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Merge branch 'dev' into snapshot_status

s4moore committed:
Merge branch 'snapshot_status' of github.com:s4moore/backintime into snapshot_status

s4moore committed:
fix --issues switch and add ability to check snapshot status on ssh drives [skip ci]

s4moore commented:
I've commited some changes (and converted pr to draft):
- fixed --issues switch
- changed default to human readable output
- added --json switch for machine readable output
- implemented checking status on ssh drives

I've been trying different layouts for the output and I think it looks neat now but I was wondering if I should add a --compact switch or something similar in case the user has lots of profiles, what do you think?

Is there any other information you want to be displayed in the profile specific view (or the overview)?

buhtz commented:
Looks good.

There is a bug in BIT that helps you somehow. :rofl: 
Your connections cause INFO logging outputs on terminal. Usually this polute the json string. But BIT has the bug that every logging output, not matter of its level, is on stderr.

So your json output and the log output is separated.

```
...
STDERR: INFO: Connecting to: unittest
STDERR: INFO: mount ssh_encfs: user@localhost:/home/user/Downloads/bit_test_ssh_encfs on /home/user/.local/share/backintime/mnt/8D19350B/mountpoint
STDERR: INFO: unmount ssh_encfs: user@localhost:/home/user/Downloads/bit_test_ssh_encfs from /home/user/.local/share/backintime/mnt/8D19350B/mountpoint
STDOUT: {
STDOUT: "Main profile": {
STDOUT: "Last snapshot": "2023-08-13 15:30:34",
STDOUT: "Successful": true
STDOUT: },
...
```

We can keep it that way. Re-write the logging machine, using Pythons in-build logging, is also on my list. But it is that big that I won't touch it yet.

Usually I would say that your feature should somehow silence the log output. But with this BIT bug there is no need. Let's keep it simple.

buhtz commented on common/backintime.py:
```diff
@@ -1145,6 +1174,161 @@ def snapshotsListPath(args):
+
+def humanPrint(dictionary, file=None, indent=0, width=-1):
+    """Print dict in human readable format."""
+    print('')
```
Check the difference of the output between `backintime status` and `backintime status --quiet`.

I am assuming that this `print()` and the one in line 1295 also should have `file=file` argument?

s4moore replied:
Ah yes, I missed that, apologies. I have updated this now.

s4moore committed:
add target file for print statements in humanPrint [skip ci]

s4moore commented:
Do you have any suggestions on other data to be included or other changes to be made? Otherwise, should I start to look at incorporating this funcitonality into the gui?

On a side note, thank you for including me in the list of contributors for the release candidate. Just a heads-up for possible future mentions, I recently changed my github username from mooresamuel to s4moore.

s4moore committed:
update from dev [skip ci]

s4moore commented:
I've made a start at incorporating this into the gui. Let me know if you think this approach works or what you think would be better. I've added a progress bar so the user can see that it's doing something in case of issues with accessing SSH drives but not sure if this is really neccessary?

![recording](https://github.com/user-attachments/assets/529f0f12-e5e1-48df-b106-f9fbb96a9a43)

s4moore committed:
change timer for flush from 5 to 1 second

s4moore committed:
create SnapshotStatus class

s4moore committed:
Reorganise SnapshotStatus and improve human print

s4moore committed:
update comment to highlight change in timer

s4moore committed:
Reintroduce issues flag

buhtz commented:
Hello Samuel,
Please excuse my late reply. I was offline because of technical reasons.

> Do you have any suggestions on other data to be included or other changes to be made?

No, not yet. I think it is OK the way it is. It is a good beginning and can be extended later.

> Otherwise, should I start to look at incorporating this funcitonality into the gui?

Having this in the GUI wasn't my use case. But I am open to it if you want to have this in the GUI.

Before coding something I would recommend that you create a mockup we can discuss. Create a new issue for this please.

I do use draw.io or Qt Creator for mock ups. But you can also draw by hand and take a photo. Every kind of visual would help.

s4moore committed:
refactor and update docstrings

s4moore committed:
refactor

s4moore committed:
merge with local branch [skip ci]

s4moore commented:
Hi Christian,

No problem, I don't feel like you were too long in replying.

I was looking at the original issue (#1018) again and saw that you wanted to be able to check the status of a drive if it was not online so I have updated the approach. It now creates a log file and updates that each time it finishes a backup so the info is available when a drive is offline. Happy to go back to the previous version if you would prefer that, just let me know.

Thanks

buhtz committed:
Merge branch 'dev' into snapshot_status

buhtz commented on common/backintime.py:
```diff
@@ -1004,6 +1038,365 @@ def snapshotsListPath(args):
         _umount(cfg)
     sys.exit(RETURN_OK)
 
+class SnapshotStatus():
```
I would suggest to move that class into its own file; eg. `common\status.py`
Additionally add that filename to the list `full_test_files` (around line 44) in `common/test/test_lint.py`. Try to keep the file free from all linters warnings.

s4moore replied:
Will do

buhtz commented on common/backintime.py:
```diff
@@ -1004,6 +1038,365 @@ def snapshotsListPath(args):
+        try:
+            with open(self.path, 'r') as f:
+                self.status = json.load(f)
+        except (FileNotFoundError, json.JSONDecodeError):
```
I would suggest to handle that two exceptions separated. This would make each error message clearer for the users.

buhtz commented on common/backintime.py:
```diff
@@ -1004,6 +1038,365 @@ def snapshotsListPath(args):
+        except (FileNotFoundError, json.JSONDecodeError):
+            logger.warning(_('Status file not found or corrupted, creating new file.'))
+            self._getAllStatus()
+        except PermissionError:
```
I see no need for this PermissionError. If there are permission problems BIT also won't be able to write a new status file.

s4moore replied:
Makes sense

buhtz commented on common/backintime.py:
```diff
@@ -1004,6 +1038,365 @@ def snapshotsListPath(args):
+            with open(self.path, 'r') as f:
+                self.status = json.load(f)
+        except (FileNotFoundError, json.JSONDecodeError):
+            logger.warning(_('Status file not found or corrupted, creating new file.'))
```
Please don't translate any messages that go into the log files.

buhtz commented on common/backintime.py:
```diff
@@ -1004,6 +1038,365 @@ def snapshotsListPath(args):
+        if not self.profile_id and self.profile:
+            self.profile_id = self.cfg.profileIntValue(self.profile)
+        self.path = \
+        os.path.expanduser('~/.local/share/backintime/snapshotstatus.json')
```
Please try to use `pathlib` and `pathlib.Path` instead of `os`.

Also the path should be modified. It is a state file not a config file. So it should go into `$XDG_STATE_HOME` (mostly `~/.local/state`). See this as an example how to construct that file path.

https://github.com/bit-team/backintime/blob/f65d503c93a3a53a8c1845c160d50bb8c83802ac/qt/statedata.py#L135-L145

buhtz commented on common/backintime.py:
```diff
@@ -1004,6 +1038,365 @@ def snapshotsListPath(args):
+        self._checkForSuccessfulSnapshot()
+
+        if success:
+            profile[_('Last run')] = {_('Timestamp'): now, _('Successful'): True,
```
Please don't translate dict keys.
For this status feature it IMHO is OK to not translate anything. It is not the GUI.

s4moore replied:
Okay

buhtz reviewed (changes requested):
Hello,
now after we released 1.5.4 we can go further with this PR and bring this to an end.

It is late in the night and I am sitting in a bumping train. So my head is not at 100%. ;)

I have not checked every detail. But it really looks good and we are on a good way.

Do I understand your overall approach right? Directly after a snapshot run you parse its log file and extract the details?

Regards,
Christian

buhtz commented:
~~I'll work on a modification of the statefile code to better support your PR.~~

~~Currently BIT has one statefile related to `backintime-qt`. A statefile for `backintime-common` is missing. In my opinion your backup-jobs-status information should be written into such a state file.~~

buhtz committed:
Merge branch 'dev' into snapshot_status

buhtz commented:
Sorry, I still mangling this PR in my head. Modifyfing the state-file code wasn't such a good 
idea as I thought before.

Can you explain me how do you distinguish if a backup job was successful or not?

s4moore commented:
Hi Christian,,

Sorry for not replying sooner, I have been quite busy the past few weeks, congratulations on another release!

Yes, after each backup, the log is parsed and the status file is updated, if it exists. If there is no status file, all profiles are checked and the status for all profiles is recorded.

A backup is counted as having been successful if there were files written without any errors. If a backup is run and no files are written (or files are written but with errors), this will be recorded as the "last run" so the user will be able to see the last time a backup job was run as well as the "last successful" backup (changes were made without any errors). OBviously if these results aren't as expected, the user could take a deeper dive through the logs as per normal.

Looking back at the code now, I realise that it is not very clear, perhaps I should refactor it? I have also just discovered an error in the logic so it does need a bit of work anyway.

I also wanted to add some tests but I struggled to get my head around the generic test classes and how to implement them and was worried that I might write tests that would pass on my machine but not pass on others due to my misunderstanding something. Should I attempt to write tests anyway? (And if so, do you have any advice or guidance on the approach I should take?)

buhtz commented:
> Yes, after each backup, the log is parsed

Mhm... But why? BIT itself somehow determines if a backup was successful or not. But I am not sure how it works. Can't you use the same mechanic? I think parsing a log file is bit to much overhead for such a simple task.

> I also wanted to add some tests

Let's skip this in the first place. I'll think about it when the code is finished.

s4moore commented:
> > Yes, after each backup, the log is parsed
> 
> Mhm... But why? BIT itself somehow determines if a backup was successful or not. But I am not sure how it works. Can't you use the same mechanic? I think parsing a log file is bit to much overhead for such a simple task.

As I understand, BIT considers a backup to be successful as long as there are no errors, even if no files are written. I thought it would be useful to know the last time that files were backed up without any errors but yes, this does add some overhead and maybe it's not really necessary.

Yes, I can use the same mechanic, I will work on another solution.

buhtz commented:
I might misunderstand something?

> As I understand, BIT considers a backup to be successful as long as there are no errors, even if no files are written.

But if no files are written, than there is no new backup.

> I thought it would be useful to know the last time that files were backed up without any errors but yes

Isn't this the same as the previous case?

s4moore commented:
Hi Christian,

Sorry, it's been a while since I worked on this. Yes, you are right. I was thinking of the log file that is written (in ~/.local/chare/backintime ) after any backup attempt (which wouldn't be an issue if I was only using the builtin mechanism).

Looking back at my code, I have actually used the builtin mechanism but also check the logs to see how many files were changed and record the error detail. I guess I was just trying, maybe too hard, to include any information which might be useful. Shall I remove that from the code then?

Edit: I also thought it would be useful for the user to be able to see the timestamp for the most recent back attempt (even if no changes were made) to confirm jobs were being run as scheduled but I guess there are other ways to do this as well

buhtz commented:
> Sorry, it's been a while since I worked on this.

No problem. It is the same on my site because I try to focus on some other things.

> Looking back at my code, I have actually used the builtin mechanism but also check the logs to see how many files were changed and record the error detail. I guess I was just trying, maybe too hard, to include any information which might be useful. Shall I remove that from the code then?

It is not bad trying it "hard". ;) Its easier to find boundaries this way. No matter that the feature might be useful it is a bit to much for the first implementation. I would say you can remove it from this PR. But don't delete it. Save it somewhere else. Maybe we can come back to it later.
I like to keep it simple in the first place to get an idea about how things might work and how the users react to it. Also it is always about maintainability.

> Edit: I also thought it would be useful for the user to be able to see the timestamp for the most recent back attempt (even if no changes were made)

That is a very nice feature! The timestamp of the last job run, even if no changes are made, should be kept. This info should stay. Distinguish between the timestamp of the last backup made and the last run (even without changes) is very useful I think.

buhtz commented:
How BIT handle snapshots logs is also new to me. You might still know it but I discovered something new to me. So this is just for your (and my) information.

In this example you see in the timeline (1.) the last backup from 11:49:45.
But clicking on Last Log View (2.) you get a log file with a timestamp from 11:50:00 (3.).
This is because I run a backup job, without any changes ("Done, no backup needed.").

![2025-04-14_11-52](https://github.com/user-attachments/assets/99bff22d-b63e-4f50-9af7-3a0cc71cefd2)

s4moore committed:
Merge branch 'dev' of github.com:bit-team/backintime into snapshot_status

s4moore committed:
Merge branch 'dev' of github.com:bit-team/backintime into simplified_status

s4moore committed:
Removed backupstatus to status.py. Change backupstatus to avoid parsing logfiles. Add status.py to test_lint.py to check all linting. Remove checks for permission errors.

s4moore committed:
linting

s4moore committed:
small change to docstring

s4moore committed:
Merge with local branch. Remove parsing log files - now uses only the inbuilt mechanism to check for successful backups. Include date of most recent log file so user can see that BIT is running as scheduled [skip ci]

s4moore committed:
Merge branch 'dev' of github.com:bit-team/backintime into snapshot_status [skip ci]

s4moore committed:
Reformat for new file structure [skip ci]

s4moore commented:
I have implemented the requested changes and tried to make the code more clear.  All seems to be working fine from my side but please let me know if there are any other changes I should make 😄

buhtz commented:
Thank you. I'll take a look. It might be that I push some commits to this PR. So please wait.

buhtz committed:
missing newline [skip ci]

buhtz committed:
reset version [skip ci]

buhtz committed:
missing newline [skip ci]

buhtz commented on common/cliarguments.py:
```diff
@@ -567,6 +568,43 @@ def _create_cmd_snapshots_path(self):
+            help=desc,
+            description=desc)
+        parser.set_defaults(func=self._cmd_func_dict[name])
+        profileGroup = parser.add_mutually_exclusive_group()
```
I like that and will adapt it for the other commands (#2132).
And now I understand why the former developers using `epilog` as a workaround.

buhtz committed:
minor mods. introduce -j and -i [skip ci]

buhtz committed:
remove debug code [skip ci]

buhtz committed:
fix linter warnings [skip ci]

buhtz commented on common/clicommands.py:
```diff
@@ -513,6 +514,12 @@ def smart_remove(args: argparse.Namespace):
 
+def backup_status(args: argparse.Namespace):
+    """Handler for CLI command 'status'"""
+    with BackupStatus(args, cfg=_get_config(args)) as status:
```
Did I suggest to design `BackupStatus` as a context manager? I have no clear opinion on that. It just feels a bit unusual. But I also don't see something wrong or bad about it. Keep it that way if you want.

s4moore replied:
No you didn't, I just thought the design was a bit more elegant. But I agree it's a bit unusual. I’ll refactor it into a standard class, and feel free to let me know if you think it works better that way

buhtz commented on common/snapshots.py:
```diff
@@ -40,6 +40,20 @@
+        finally:
+            with BackupStatus(cfg=self.config) as status:
+                status.update_status()
+    return wrapper
```
I am not a friend of decorators but would accept them for good reasons.

I might do understand why you added `@_update_backup_status_after_takeSnapshot`. You want to prevent multiple calls of `BackupStatus.update_status()` in `Snapshots.takeSnapshot()`, right? IMHO the decorator makes code harder to read and understand, because it is somehow implicit.

I would prefer a direct call to `BackupStatus.update_status()` at the end of a backup job. I am aware that the code in `Snapshots.takeSnapshot()` is a smelling mess and there is not only "one end" but several lines with `return`. Someday this method will get a refactoring, but not today. Until then I suggest to check each `return` statement in this method and simply add `BackupStatus.update_status()` in front of it.

s4moore replied:
I've now changed this so backupstatus is called from snapshots.Snapshots.backup straight after a backup attempt.

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,331 @@
+    """
+    def __init__(self,
+                 args: argparse.Namespace = None,
+                 cfg: config.Config = None):
```
I would prefer to decouple the class from `argparse` module.
My quick n dirty suggestion:

```suggestion
    def __init__(self,
                 cfg: config.Config = None,
                 profile: str = None,
                 profile_id: int = None,
                 with_issues: bool,  #  default value will come from the args object
                 as_json: bool):
```

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,331 @@
+                 cfg: config.Config = None):
+
+        self.profile = args.profile if args else None
+        self.profile_id = args.profile_id if args else None
```
`self.profile` and `self.profile_id` seems somehow redundant to me. Maybe just keep `self.profile_id` and request the name from the `cfg` object if you need it?

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,331 @@
+
+    def _create_profile_status(self):
+        """
+        Create status entry for profile specified by self.profile_id.
```
Can you please review your docstrings in the methods related to dict and output handling. To me as a reader it is not clear what they are doing, without reading the code.

e.g. "status entry": Entry in what? The dict?

My intention behind this, is that I want to write unit tests for your class.

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,331 @@
+
+        return max_len
+
+    def _human(self, dic, indent=0, width=None):
```
This `_human()` method, as an example, seems Independent from `class BackupStatus`. It does not use its instance attributes or something else. With having unit tests writing in mind, I would prefer to decouple methods like this from the class instance.

One option is to make them a `@staticmethod` or `@classmethod`. Or you simply remove them from the class and make them a function in the module `status`.

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,331 @@
+
+        return '\n'.join(human)
+
+    def _file_path(self):
+        xdg_state = os.environ.get('XDG_STATE_HOME', None)
+        if xdg_state:
+            xdg_state = Path(xdg_state)
+        else:
+            xdg_state = Path.home() / '.local' / 'state'
+
+        return xdg_state / 'backupstatus.json'
```
```suggestion
    def _file_path(self):
        xdg_state = os.environ.get(
            'XDG_STATE_HOME',
            Path.home() / '.local' / 'state')

        return Path(xdg_state) / bitbase.BINARY_NAME_CLI / 'backupstatus.json'
```

`bitbase.BINARY_NAME_CLI` is "backintime". Don't forget to `import bitbase` on top of the file.

I know it looks akward, because in case `XDG_STATE_HOME` is not present, the default value is of type `Path` and because of that, in the `return` statement, `Path()` will be initialized with that default `Path` instance. But it works.

s4moore replied:
is this a new location? On my machine, the state files are in ~/.local/state

buhtz replied:
Ah, I see. You are right, currently there is not dedicated sub-folder for BIT related state files. And I think we don't need it.

In this case I would recommend to modify the name of your state file but not the path.

    Path.home(xdg_state) / 'backintime-status.json'

You can mke the name longer and more informative if you want, e.g. `backintime-backup-status.json`. That files are not noticed or read by human beings in most cases.

buhtz reviewed (changes requested):
I'm a bit worried that I'm annoying you with all my comments and suggestions.

The primary aspects are the decorator and the failing tests. The rest is "high level" pedantic code quality stuff, like the cherry on top of the ice cream. 😄 🍨 🍒 

Beside my inline-comments:

1. There where some linter warnings, I fixed. Please install `ruff` (maybe `flake8`) and `pylint` on your local machine to see them while running tests.
2. The test `test/test_takeSnapshot.py::TakeSSH::test_error` fails. Also some other tests, but maybe for the same reason. And I don't understand why but I think the decorator is somehow involved.

s4moore commented:
> I'm a bit worried that I'm annoying you with all my comments and suggestions.

Haha not at all, I'm a bit worried that I'm annoying you with writing code that isn't hitting the mark! 🙃 

> 1. There where some linter warnings, I fixed. Please install `ruff` (maybe `flake8`) and `pylint` on your local machine to see them while running tests.

This is strange, I have run checks with all of them and didn't get any errors...
```bash
test@test-VirtualBox:~/backintime/common$ pytest test/test_lint.py
================================================= test session starts ==================================================
platform linux -- Python 3.12.3, pytest-7.4.4, pluggy-1.4.0
rootdir: /home/test/backintime/common
plugins: pyfakefs-5.7.4
collected 6 items

test/test_lint.py ......                                                                                         [100%]

============================================= 6 passed in 72.69s (0:01:12) =============================================
test@test-VirtualBox:~/backintime/common$ pylint status.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

test@test-VirtualBox:~/backintime/common$ ruff check status.py
All checks passed!
test@test-VirtualBox:~/backintime/common$ flake8 status.py
test@test-VirtualBox:~/backintime/common$
```
Do you have any idea what I'm doing wrong so I can try to avoid this in the future?

> 2. The test `test/test_takeSnapshot.py::TakeSSH::test_error` fails. Also some other tests, but maybe for the same reason. And I don't understand why but I think the decorator is somehow involved.

Okay, I will remove the decorator. I was only trying to avoid having the same call multiple times but I see from your comments that you're okay with that for now so it's an easy fix 😄

buhtz commented:
> I'm a bit worried that I'm annoying you with writing code that isn't hitting the mark! 🙃

That is the usual case. Even with myself. Just look at the PRs I am opening. There are often a lot of iterations until I merge them.

> This is strange, I have run checks with all of them and didn't get any errors...
> test/test_lint.py ...... [100%]

Try with "-v" switch. The linter tests are skipped by default (only on local machines but not TravisCI) if the linters are not available in the system. The tests don't fail. My intention was not to annoy fresh contributors with pedantic linter warnings. 😄 

But now all linter warnings are fixed by me. This might be the reaons why the tests passing?

> > 2. The test `test/test_takeSnapshot.py::TakeSSH::test_error` fails. Also some other tests, but maybe for the same reason. And I don't understand why but I think the decorator is somehow involved.
> 
> Okay, I will remove the decorator. I was only trying to avoid having the same call multiple times but I see from your comments that you're okay with that for now so it's an easy fix 😄

But I think removing the decorator won't make the tests pass. There seems to be more behind.

buhtz commented:
btw: Not sure if you read the mailing list. Because you are currently involved with the CLI interface, you might have an opinion about [this question](https://mail.python.org/archives/list/bit-dev@python.org/thread/VJ72667IFAIB3LO5S3W7YRCET7TTQVPD/)?

s4moore committed:
Merge remote-tracking branch 'upstream/dev' into snapshot_status [skip ci]

buhtz commented:
Let me know If I can be of assistance solving the merge conflicts. I was the one modifying those files.

s4moore committed:
Merge remote-tracking branch 'upstream/dev' into snapshot_status

s4moore committed:
Change backup status from a context manager to a normal class. Add more docstrings and comments to make the code more clear. Remove methods not relying on class instance to functions outside of the class.[skip ci]

s4moore commented:
Sorry this has taken a while, I've got quite a bit going on right now.

I've changed to a standard class, removed the decorator, and made the other changes requested. All tests seem are passing, including ssh_tools, so hopefully you get the same results on your side. I've also added more docstrings and comments so hopefully all is more clear now?

buhtz commented:
Hello Samuel,

thank you very much for the modifications. I will look deeper into it later.

> Sorry this has taken a while, I've got quite a bit going on right now.

No need to say sorry. Really! Open source shouldn't be about saying and/or feeling sorry or explaining why something took longer than expected. This is unhealthy. Don't push yourself.
Yes, open source is also about responsibility but not only for the project and their users but also for yourself. Open source should be fun, not pressure! ❤️ 

Relax. 😃 

Regards,
Christian

buhtz committed:
rename backup_status() to status()

buhtz committed:
rearranged parsers

buhtz committed:
[skip ci]

buhtz commented:
I'll do some more refining tonight or tomorrow.

buhtz committed:
x

buhtz committed:
[skip ci]

buhtz committed:
[skip ci]

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+            dict:
+                        Status data for the profile.
+        """
+        original_profile_id = self.cfg.currentProfile()
```
This method now remembers the current profile itself and set it back when it ends.

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+            # Get the last run timestamp
+            if timestamp is not None:
+                # If a timestamp is provided, use it for the last run
+                last_log_ts = timestamp.strftime("%Y-%m-%d %H:%M:%S")
```
I would prefer `%X %x` as timestamp template. It is the current locals time format.

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+
+            # If timestamps for most recent and last successful backup
+            # are the same, then the last backup was successful.
+            if last_backup and last_success_ts == last_backup_ts:
```
I am not convinced about this solution and assumption. A backup can be complete but can have errors. In this case you will see "(WITH ERRORS)" behind the the timestamp name in the timeline widget in the main window. I am not sure about the technical details.

For this PR I would keep it simple. Just show the last run and last backup timestamp. Not more.

Not this

```
share_SSH:
  Last Run      : 2025-05-26 21:01:46
  Last Backup:
    Status      : Success
    Completed At: 2025-05-26 21:00:47

```

But this
```
share_SSH:
  Last Run      : 2025-05-26 21:01:46
  Last Backup   : 2025-05-26 21:00:47
```

We can upgrade the status details later.

s4moore replied:
Just to clarify, you want only the most recent backup (even if it has errors)? And also to remove the extra information (backup mode and paths) when viewing the status for a specific profile?

buhtz replied:
No.

```
share_SSH:
  Last Run      : 2025-05-26 21:01:46
  Last Backup   : 2025-05-26 21:00:47
```
Backup mode is there. Keep it.
Backup path? Was it there before? You can keep that.

> only the most recent backup

No, the most recent backup ("Last Run") and the last run ("Last Run" I think).
The point is currently you don't know yet if the last backup has errors or not. For a later PR it would be nice to have this information. But it need some more investigation how to get that information. Your approach by comparing the timestamps is IMHO not valid (trustful?) enough.

buhtz reviewed (changes requested):
OK, this looks nice.
Check careful my changes and/or test them. I might have broken something. 😆 

I modified the handling of the profile_id in BackupStatus class.
I also tweaked `_human()` a bit.

I really love recursion. But I am not sure if we need it to print a simple two level dictionary. What do you think? The code is hard to read for outsiders.

Despite of that I would say this is nearly finished. What do you think?

Regads,
Christian

buhtz committed:
fix linter warnings

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+
+        except json.JSONDecodeError:
+            logger.warning('Error reading status file, creating new file.')
+            self._create_status_file()
```
```suggestion
        except (FileNotFoundError, json.JSONDecodeError) as exc:
            logger.warning('Problems while reading status file. '
                           f'Creating new file. Error was {exc}')
            self._create_status_file()
```

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+        """
+        try:
+            with open(_status_file_path(), 'r', encoding='utf-8') as f:
+                self.status = json.load(f)
```
One letter variable names are not a good choice.

```suggestion
            with open(_status_file_path(), 'r', encoding='utf-8') as handle:
                self.status = json.load(handle)
```

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+        """
+        try:
+            with open(_status_file_path(), 'w', encoding='utf-8') as f:
+                json.dump(self.status, f, indent=2)
```
Why do you use `indent` here? Does human beings need to read the file?
```suggestion
            with open(_status_file_path(), 'w', encoding='utf-8') as handle:
                json.dump(self.status, handle, indent=2)
```

buhtz commented on common/status.py:
```diff
@@ -0,0 +1,411 @@
+
+    timestamp_ts = datetime.fromtimestamp(timestamp)
+
+    return timestamp_ts.strftime("%Y-%m-%d %H:%M:%S")
```
```suggestion
    return timestamp_ts.strftime("%x %X")
```

buhtz committed:
fix regex related test

buhtz committed:
x

buhtz committed:
x

buhtz committed:
Merge branch 'dev' into snapshot_status

s4moore committed:
Update common/status.py

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/status.py

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore committed:
Update common/status.py

Co-authored-by: Christian Buhtz <c.buhtz@posteo.jp>

s4moore commented:
> OK, this looks nice. Check careful my changes and/or test them. I might have broken something. 😆
 
Sure thing 😁 

> I modified the handling of the profile_id in BackupStatus class. I also tweaked `_human()` a bit.
> 
> I really love recursion. But I am not sure if we need it to print a simple two level dictionary. What do you think? The code is hard to read for outsiders.

Yes, when I started writing this, I had no idea how much information would be displayed or if more would be added over time so thought I was "future proofing" by using a recursive approach. Now that the requirements are more clear, it does seem quite unneccessary so I will make a simpler solution 👌 

> Despite of that I would say this is nearly finished. What do you think?

It feels like we're getting there 😁

buhtz commented:
> > Despite of that I would say this is nearly finished. What do you think?
> 
> It feels like we're getting there 😁

:)

Is this a "ready to merge" or should I wait for more to come?

buhtz committed:
Merge branch 'dev' into snapshot_status

buhtz committed:
Merge branch 'dev' into snapshot_status

s4moore committed:
removed check for backups with errors, simplified printing, removed issues flag [skip ci]

s4moore commented:
I have removed the checking for successful backups based on date, simplified the human readable output, removed the indent frmo json, and removed the issues flag (no longer relevant without checking for backups with errors). Linting passes on my machine with pylint, ruff, and flake8 (with --verbose and -v flags) but have skipped ci so you can check from your side if you like. Please let me know if you would like any other changes.

buhtz reviewed (approved):
Hello Samuel,
On behalf of the team, thank you for your contribution and taking
the time to improve Back In Time. We appreciate it.

This looks fine now and is IMHO ready to merge.

But if you see a way one final thing should be accomplished. The man page need an update. You find it in `doc/manpages/backintime.1.adoc`. The format is "asciidoc", kind of Markdown and shouldn't be that hard to edit.

Would be great if you could add this, too.

Regards.
Christian

**EDIT**: And a changelog entry is missing. But I can add it if you want.

buhtz commented:
FYI: Writing it myself.

**EDIT**: OK, I updated the man page and added a rudimentary section in the manual.

buhtz committed:
Merge branch 'dev' into snapshot_status

buhtz committed:
x

buhtz committed:
x

buhtz committed:
changelog [skip ci]

buhtz commented:
Hello Samuel,
regarding this PR, can you please have a look in my comment in another issue.

https://github.com/bit-team/backintime/issues/2296#issuecomment-3522722124

Your "--status" is involved and I have no idea how to solve it. It need to be fixed until the next release (1.6.0) around spring 2026. Otherwise I need to disable the "--status" feature for a while until we improve our understanding and control over this mounting-stuff in BIT.

Regards,
Christian

- #2321
